File: Language/C#/Book/tmp/LINQ实战/Chapter12.VB/LinqToAmazon/Query.vb
' The code in this file comes from Matt Warren's series of blog posts on how to build a LINQ provider
' http://blogs.msdn.com/mattwar/archive/2007/08/09/linq-building-an-iqueryable-provider-part-i.aspx
' Converted to VB

Imports System.Linq.Expressions

Public Class Query(Of T)
  Implements IQueryable(Of T)
  Implements IQueryable
  Implements IEnumerable(Of T)
  Implements IEnumerable
  Implements IOrderedQueryable(Of T)
  Implements IOrderedQueryable
  Private m_provider As QueryProvider
  Private m_expression As Expression

  Public Sub New(ByVal provider As QueryProvider)
    If provider Is Nothing Then
      Throw New ArgumentNullException("provider")
    End If
    Me.m_provider = provider
    Me.m_expression = Expression.Constant(Me)
  End Sub

  Public Sub New(ByVal provider As QueryProvider, ByVal expression As Expression)
    If provider Is Nothing Then
      Throw New ArgumentNullException("provider")
    End If
    If expression Is Nothing Then
      Throw New ArgumentNullException("expression")
    End If
    If Not GetType(IQueryable(Of T)).IsAssignableFrom(expression.Type) Then
      Throw New ArgumentOutOfRangeException("expression")
    End If
    Me.m_provider = provider
    Me.m_expression = expression
  End Sub

  Private ReadOnly Property Expression() As Expression Implements IQueryable.Expression
    Get
      Return Me.m_expression
    End Get
  End Property

  Private ReadOnly Property ElementType() As Type Implements IQueryable.ElementType
    Get
      Return GetType(T)
    End Get
  End Property

  Private ReadOnly Property Provider() As IQueryProvider Implements IQueryable.Provider
    Get
      Return Me.m_provider
    End Get
  End Property

  Public Function GetEnumerator() As IEnumerator(Of T) Implements IEnumerable(Of T).GetEnumerator
    Return DirectCast(Me.m_provider.Execute(Me.m_expression), IEnumerable(Of T)).GetEnumerator()
  End Function

  Private Function GetEnumeratorNonGeneric() As IEnumerator Implements IEnumerable.GetEnumerator
    Return DirectCast(Me.m_provider.Execute(Me.m_expression), IEnumerable).GetEnumerator()
  End Function

  Public Overloads Overrides Function ToString() As String
    Return Me.m_provider.GetQueryText(Me.m_expression)
  End Function
End Class

File: Language/C#/Book/LINQ实战/第12章 扩展LINQ/Chapter12.CSharp/LinqToAmazon/Query.cs
// The code in this file comes from Matt Warren's series of blog posts on how to build a LINQ provider
// http://blogs.msdn.com/mattwar/archive/2007/08/09/linq-building-an-iqueryable-provider-part-i.aspx

using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using System.Reflection;
using System.Text;

namespace LinqInAction.Chapter12.LinqToAmazon
{
    public class Query<T> :
      IQueryable<T>, IQueryable,
      IEnumerable<T>, IEnumerable,
      IOrderedQueryable<T>, IOrderedQueryable
    {
        QueryProvider provider;
        Expression expression;

        public Query(QueryProvider provider)
        {
            if (provider == null)
            {
                throw new ArgumentNullException("provider");
            }
            this.provider = provider;
            this.expression = Expression.Constant(this);
        }

        public Query(QueryProvider provider, Expression expression)
        {
            if (provider == null)
            {
                throw new ArgumentNullException("provider");
            }
            if (expression == null)
            {
                throw new ArgumentNullException("expression");
            }
            if (!typeof(IQueryable<T>).IsAssignableFrom(expression.Type))
            {
                throw new ArgumentOutOfRangeException("expression");
            }
            this.provider = provider;
            this.expression = expression;
        }

        Expression IQueryable.Expression
        {
            get { return this.expression; }
        }

        Type IQueryable.ElementType
        {
            get { return typeof(T); }
        }

        IQueryProvider IQueryable.Provider
        {
            get { return this.provider; }
        }

        public IEnumerator<T> GetEnumerator()
        {
            return ((IEnumerable<T>)this.provider.Execute(this.expression)).GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return ((IEnumerable)this.provider.Execute(this.expression)).GetEnumerator();
        }

        public override string ToString()
        {
            return this.provider.GetQueryText(this.expression);
        }
    }
}